bazel_tools/client_server/runner_with_port_file/Main.hs

```diff
@@ -0,0 +1,71 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-% LANGUAGE OverloadedStrings %-}
+
+module Main(main) where
+
+import Control.Concurrent
+import Control.Exception.Safe
+import Control.Monad (forM, forM_)
+import Control.Monad.Loops (untilJust)
+import Data.List.Split (splitOn)
+import Network.Socket
+import System.Directory (removeFile)
+import System.Environment
+import System.IO.Temp (emptySystemTempFile)
+import System.Process
+import qualified Data.Text as Text
+
+main :: IO ()
+main = do
+  [clientExe, clientArgs, serverExe, serverArgs, _] <- getArgs
+  let splitArgs = filter (/= "") . splitOn " "
+  (portFile, serverArgs') <- generatePortFile (splitArgs serverArgs)
+  let serverProc = proc serverExe serverArgs'
+  withCreateProcess serverProc $ \_stdin _stdout _stderr _ph -> do
+    maybePort <- forM portFile $ \file -> do
+      putStrLn $ "Waiting for the port in the file \"" <> file <> "\" to open..."
+      waitForConnectionInPortFile (threadDelay 500000) file
+    let clientArgs' = case maybePort of
+                        Nothing -> splitArgs clientArgs
+                        Just port -> map (interpolateVariable "PORT" (show port)) (splitArgs clientArgs)
+    callProcess clientExe clientArgs'
+  forM_ portFile removeFile
```

Comment:
this is not exception safe. If your code fails before this the file might persist. I recommend to use `withTempDir` to create a directory and then chose a file path in there. Then you also don’t have to worry about deleting the temp file you just created. We use that a fair bit throughout the codebase, e.g., https://github.com/digital-asset/daml/blob/0c1878530b5341e52f1eb5f107ce109ce5953f19/daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs#L103

Reply:
That makes sense; I'll do that. Thanks.